### PR TITLE
[Mechanic] Hull / Shield Repair Limits and Wear

### PR DESCRIPTION
**Pull request by Azure3141**
**Feature**

This PR addresses the feature requested by the voices in my head

## Summary
This PR adds two sets of attributes for placing limits on how much ships can regenerate their shields or hull.

`nonregenerative shields` and `nonrepairable hull` remove a fraction of a ship's maximum health that can be regenerated without landing at a suitable spaceport. For example, an outfit with `"nonrepairable hull" 0.1` would prevent a ship from repairing its hull beyond 90% of its maximum. Two of such outfits installed would reduce the limit to 80%, and so on.

`shield generation wear` and `hull repair wear` increment a counter by the attribute's value every frame the shields or hull are being regenerated. This counter subtracts from the maximum shields or hull that can be regenerated in flight, and is reset upon landing at a spaceport with shield or hull repair facilities. The wear counters are independent for shields and hull. For example, a shield generator with `"shield generation wear" 0.5` would remove 0.5 points of maximum shielding from a ship every frame it was active. If said ship had a shield capacity of 1000, lost 200 points of shields, then spent 100 frames regenerating, the new maximum shield strength would be 950. Further attrition would steadily lower this as more damage was taken and then regenerated.

Multipliers for wear generation are also included.

## Screenshots
N/A unless you want to see pictures of a ship missing some HP.

## Usage examples
If you had a thing against human shields for some reason, you could do this:
```outfit "D94-YV Shield Generator"
	category "Systems"
	cost 200000
	thumbnail "outfit/huge shield"
	"mass" 60
	"outfit space" -60
	"shield generation" 1.31
	"shield energy" 1.31
	"nonregenerative shields" 0.5
```

This would prevent a ship from regenerating its shields beyond 50% of their maximum. Ships would launch with fresh shields at full capacity, but if they took damage they would be unable to repair it until their shields dropped below 50%, which would also be the maximum that could be repaired to.

Alternatively, if you really thought the Smew was the most OP and unbalanced Remnant warship, you could do something like this:

```ship "Smew"
	sprite "ship/smew"
	thumbnail "thumbnail/smew"
	attributes
		category "Fighter"
		licenses
			Remnant
		"cost" 971000
		"shields" 1200
		"hull" 2800
		"shield generation" 0.48
		"shield energy" 0.32
		"hull repair rate" 0.9
		"hull energy" 0.58
		[other attributes]
		"hull repair wear" 0.45
```
As the Smew regenerates 0.9 hull in a frame, it would also lose maximum hull health at the rate of 0.45 points per frame. This effectively means the Smew would only be able to regenerate half the damage it took.

Hopefully this would be able to nerf its terrifying combat capabilities.

## Testing Done
Made some ships, stuck outfits on them, shot them and kept track of the shield repair.

## Save File
N/A

## Performance Impact
Probably none.

### CODE REVIEW

File: source/Ship.cpp [hunk 1]
```diff
@@ -2607,13 +2618,32 @@ double Ship::MaxShields() const
...
+{
+	double shieldLimit = attributes.Get("nonregenerative shields");
+
+
+	return 1 - shieldLimit;
+}
+
+
+
+double Ship::HullRepairLimit() const
```
Comment (RisingLeaf) [inline source/Ship.cpp:2639]: Seems unnecessary, just name the attributes limit and there is no need for this conversion

File: source/Ship.cpp [hunk 2]
```diff
@@ -2413,10 +2414,20 @@ void Ship::Recharge(int rechargeType, bool hireCrew)
 	pilotError = 0;
 	pilotOkay = 0;
 
-	if((rechargeType & Port::RechargeType::Shields) || attributes.Get("shield generation"))
+	if(attributes.Get("shield generation"))
+		shields = max(shields, MaxShields() * ShieldRepairLimit());
+	if(attributes.Get("hull repair rate"))
+		hull = max(hull, MaxHull() * HullRepairLimit());
```
Comment (RisingLeaf) [inline source/Ship.cpp:2420]: this wont cause a sudden drop?
I mean if you got almost full hull, it tries to repair and then it drops to the limit?

File: source/Ship.cpp (hunk 1 above)
Comment (RisingLeaf) [reply source/Ship.cpp]: No no, what I mean is rename the attribute so you dont need to calc

Comment (RisingLeaf) [reply source/Ship.cpp]: Ok

File: source/Ship.cpp (hunk 2 above)
Comment (RisingLeaf) [reply source/Ship.cpp]: Makes sense:+1:

**Review by RisingLeaf — APPROVED** [review]

**Comment by demolish238** [comment]
I don't think this makes any sense for shields, but for hull it seems logical

**Comment by RisingLeaf** [comment]
Why not? We have no model of how shields work in ES so it depends on the explanation.

**Comment by demolish238** [comment]
Shields are what are designed specifically to take a continual beating, if they got worse at that over time and needed constant tune ups that would be a design flaw, while hull repair is something that would rarely be needed in a constant basis, often just as an emergency repair to get straight back into battle. There could be exceptions, but that's how the current two systems work in game in my view.

**Review by ziproot — APPROVED** [review]

**Review by OcelotWalrus — APPROVED** [review]

**Comment by quyykk** [comment]
Idk I'm not really a fan. Seems to overcomplicate the shields/hull mechanic for no good reason. But I'll defer to others.

**Review by petervdmeer — CHANGES_REQUESTED** [review]
I agree with quyykk that this PR (as-is) seriously complicates the hull and shield repair mechanics. I had to read the descriptions 3 times to understand what we actually want to do here, which is more than what we can expect from a casual gamer (the target-audience for endless sky).

The feature itself sounds pretty cool though; dynamic limits on how much and how often you can re-generate hull (and shields) in-flight.

This feature might benefit from more clear keywords:
- If I see "nonrepairable hull", then I get the feeling that I'm dealing with a spaceship that has "ablative armour" on it's hull. When it's gone, then it's gone, also not repairable in spaceports. (And I would really like to see such a feature for some "cool but limited use" ships, but that is not this PR.)
- I feel that the word "inflight" needs to be in the keywords, since the full hull and shields can be repaired in spaceport.


Maybe we can simplify the feature a bit:
- Remove the multipliers (it is already complex as-is, and I don't see much value in having those multipliers)
- Remove the absolute maximums (the "wear" maximums are much more fun)
- Remove this mechanism for shields (we don't need shields and hull to be the same, and this feature seems to feel more natural for material hulls than for energy shields). Shields were originally always intended to be regenerate-able in-flight, I feel that we shouldn't start limiting that mechanic. (Or at least not by absolute limits.)
- Don't use the word "wear", since it it not permanent wear (because full repairs can be done in spaceport, the hull and shields are not wearing out, there is just a limit on what you can repair yourself)
- Keep only the dynamic hull limit, and call it something like "inflight hull repair efficiency" or "max inflight hull repair", with a description like: "Only this percentage of damaged hull sections can be repaired in-flight. The percentage is not an absolute maximum, but applies to damaged sections each time a ship takes damage. Full repairs typically require landing on a spaceport."
- Let the mechanic work on repaired hull points, so an efficiency of 0.9 would mean, for every 10 points of hull restored, the maximum moves down by 1. (Not based on frames, casual players don't know frames.)

Visual markers in the UI in the hull indicators (to show the limits of repair) would also be nice. (But can be kept for a follow-up PR.)
A text message in the UI might also be helpful in the meantime. (Your ship was repaired inflight, but full repairs requires landing at a spaceport.)

File: source/Ship.h [hunk 3]
```diff
@@ -615,6 +615,8 @@ class Ship : public Body, public std::enable_shared_from_this<Ship> {
 	// Various energy levels:
 	double shields = 0.;
 	double hull = 0.;
+	double shieldsWear = 0.;
+	double hullWear = 0.;
```
Comment (petervdmeer) [inline source/Ship.h:619]: I get the impression that those values also need to be saved to savegames (for ships that are not at a spaceport when a game is saved.)

File: data/_ui/tooltips.txt [hunk 4]
```diff
@@ -282,6 +282,15 @@ tip "hull heat:"
...
 	`Modifies the heat produced when repairing the hull by the given factor.`
 
+tip "hull repair wear:"
+	`Temporary maximum hull lost per second when repairing the hull.`
+
+tip "hull wear multiplier:"
+	`Modifies the temporary maximum hull lost when repairing the hull by the given factor.`
+
+tip "nonrepairable hull"
+	`The percentage of the ship's hull that cannot be repaired.`
```
[the remark this answers is not in the document]
Comment (OcelotWalrus) [reply data/_ui/tooltips.txt:292]: Maybe `The percentage of the ship's hull that cannot be repaired in-flight.`?

File: data/_ui/tooltips.txt [hunk 5]
```diff
@@ -282,6 +282,15 @@ tip "hull heat:"
 tip "hull heat multiplier:"
 	`Modifies the heat produced when repairing the hull by the given factor.`
 
+tip "hull repair wear:"
+	`Temporary maximum hull lost per second when repairing the hull.`
```
[the remark this answers is not in the document]
Comment (OcelotWalrus) [reply data/_ui/tooltips.txt:286]: I propose `Decreases hull repair efficiency in-flight`. But perhaps it doesn't give enough explication?

**Comment by Zitchas** [comment]
Would it be possible to have the  "can't regenerate/repair back up beyond X" set to trigger on reaching a specific condition?

For instance, a common thing I can think of would be that most ships can repair/regen 100% most of the time; but if they take *massive* amounts of damage (say, over 50%) then the restriction cuts in.

From a storytelling/lore perspective, there's a bit difference between a ship getting pounded to within an inch of its life and a ship getting a scratch from a single missile across the bow. The former should be really hard to get back to full health (maybe even impossible without a shipyard, never mind a planet); whereas the latter should be very easy. 

Likewise, as another idea, it'd be interesting to have quantity limits. As in being able to say that the repair system can handle doing up to X amount of hp worth of repairs, and after that it is exhausted and doesn't work anymore. (likewise the alternative, that it can repair up to full health until it reaches its limit; and then after that it can only repair up to whatever its limit is)

Overall, I like the addition; however in some ways I feel like this is just tiptoeing around the idea of splitting hull into armor and hull. Armor is the stuff on the outside that can be repaired comparatively easily. Hull is the stuff on the inside that represents structure, electronics, wiring, fuel distribution, environmental systems, and the million and one other "moving parts" (note. Not necessarily actually moving, just active in the sense that they're actually doing things, not just being passive plating to take hits) that are in all likelihood far more difficult to repair. Amusingly, though, I can also see the opposite being the case. Very easy to replace some wiring and tubing; Carrying some replacements of those (or even manufacturing some of it) is something that I could see a ship being able to do without unduly infringing on believability. But carrying big thick sheets of armor plating or inventories of reactive sheathing? That's very bulky, and would be akin to carrying around most of another ship's worth of mass just to be able to fix the armor. It depends a lot on the story behind the various faction's systems, and either way, the best implementation of that is making "hull" a separate thing from the "armor", with separate repair values.

File: data/_ui/tooltips.txt (hunk 4 above)
[the remark this answers is not in the document]
Comment (petervdmeer) [reply data/_ui/tooltips.txt:292]: That would indeed be a better description.

**Comment by petervdmeer** [comment]
> Would it be possible to have the "can't regenerate/repair back up beyond X" set to trigger on reaching a specific condition?

The `"hull repair wear"` as proposed here seems to be doing that in quite an elegant way, because for every X (say 10) points of damage, only Y (say 9) points get repaired in flight;
- So if a ship sustains only minor damage (say 20 points), then 18 of those points get repaired.
- If a ship has massive damage (say 200 points), then 180 points get repaired, and 20 not.
- And this applies each time a ship gets damaged, so for example a ship getting damaged 20x for 20 points, will only have 20x18 points repaired. (Thus 40 points of damage not repaired after getting all those damages.

> For instance, a common thing I can think of would be that most ships can repair/regen 100% most of the time; but if they take _massive_ amounts of damage (say, over 50%) then the restriction cuts in.

The nice thing about the mechanism as proposed in this PR is that the restriction is always active, but the impact of minor damage is just minor, while the impact of many minor damages and the impact of major damage is significant.
 
> From a storytelling/lore perspective, there's a bit difference between a ship getting pounded to within an inch of its life and a ship getting a scratch from a single missile across the bow. The former should be really hard to get back to full health (maybe even impossible without a shipyard, never mind a planet); whereas the latter should be very easy.

Agreed, but the latter doesn't need to be perfect; emergency repairs while a vehicle is active is typically weaker than the repairs done in an official repair shop.

> Likewise, as another idea, it'd be interesting to have quantity limits. As in being able to say that the repair system can handle doing up to X amount of hp worth of repairs, and after that it is exhausted and doesn't work anymore. (likewise the alternative, that it can repair up to full health until it reaches its limit; and then after that it can only repair up to whatever its limit is)

The `"hull repair wear"` as proposed actually creates this limit, but with a nice sliding scale. So it is not like "having X amount worth of repairs, and then it's gone", but it is like "for every piece of damage my ship gets a reduced upper limit of hull strength". I feel that this is much more fun, since it is not a "boolean" type of limit, but the risk of staying in some combat-rich environment just slowly increases. (Should I try to capture another ship, or does my ship now really get to weak?)

> Overall, I like the addition; however in some ways I feel like this is just tiptoeing around the idea of splitting hull into armor and hull.

Splitting armor and hull is also a nice idea, but that is not this PR. This PR is actually a pretty elegant nerf to hull repair; it doesn't disable hull repair completely, but it allows ships to degrade in max strength based on damage taken. (And it does indirectly put a hard-limit on the amount of repairs that can be done.)

File: data/_ui/tooltips.txt (hunk 5 above)
[the remark this answers is not in the document]
Comment (petervdmeer) [reply data/_ui/tooltips.txt:286]: How about `"Decreased hull repair efficiency in-flight. Only this percentage of damaged hull sections can be repaired in-flight. The percentage is not an absolute maximum, but applies to damaged sections each time a ship takes damage."` ?

[the remark this answers is not in the document]
Comment (OcelotWalrus) [reply data/_ui/tooltips.txt:286]: It does give more info, but maybe too much in my opinion.

**Comment by petervdmeer** [comment]
> > Remove the multipliers (it is already complex as-is, and I don't see much value in having those multipliers)
> 
> Multipliers allow you to have outfits that modify the behavior of shield/hull repair outfits. The Shield Refactor Module is a good example of how this can lend to interesting builds, and this could be used to create something like a hull repair outfit that's really good but uses both hull wear and a multiplier to discourage stacking them.

I know, they add new posibilities, but they also add complexity. I still feel that the added complexity doesn't outweigh the added benefits and I strongly recommend to keep the multipliers out of this PR and focus on the core mechanic first.

> > Remove the absolute maximums (the "wear" maximums are much more fun)
> 
> You might like the wear system more, but I think repair maximums are still a useful feature for content creators working with less advanced species to create limited repair outfits without utterly crippling them by giving them large wear penalties. The use case I had in mind was an Avgi hull repair bay that would require a large crew complement and only work to repair hull up to a maximum of 50% or so. This would make taking hull damage permanent to a degree, but allow players to still recover some of their hull if they could escape just before death. Only having wear available wouldn't really allow this behavior, as it lends itself more towards nerfing long-term attrition more than enabling repair outfits that can help save you from rapid death while not letting you completely recover.

Maybe it would work to give those Avgi a "hull wear" attribute of 30%. The first time they get seriously damaged allows repairs up to 70%, but the second time they get serious damage only allows repairs up to 49%. (0.7*0.7 = 0.49), and the third time they get this seriously damaged only allows (0.7 * 0.7 * 0.7 = 0.34%). It is a less advanced species, so it might make sense for those species to also have this limit on attrition.

> > Remove this mechanism for shields (we don't need shields and hull to be the same, and this feature seems to feel more natural for material hulls than for energy shields). Shields were originally always intended to be regenerate-able in-flight, I feel that we shouldn't start limiting that mechanic. (Or at least not by absolute limits.)
> 
> I disagree with this. For the same reason that we have attributes allowing shield generators to produce disruption damage, if this system exists for hull the same functionality should be available for shields, even if it's only used in plugins.

We don't build stuff to only be used in plugins. It is either good enough for use in the main game, or it shouldn't be in. And the main argument here still is; Can we explain this to a casual gamer? (The target audience of ES.) This doesn't only apply to casual gamers, our content creators also need to understand every mechanic we put into the game.

The basic premise of the game currently is; "Shields always restore to full, when the resources are available." This is very easy to explain.

> > Don't use the word "wear", since it it not permanent wear (because full repairs can be done in spaceport, the hull and shields are not wearing out, there is just a limit on what you can repair yourself)
> 
> What word would you suggest? To me wear suggests something that can't be easily repaired in the field but which can be addressed by more involved maintenance / replacement of parts.

English isn't my native language, so maybe I understand the meaning of the word wrong. But I always thought "wear" to mean something getting worse over time due to usage. This is not the hull getting worse over time due to usage, but the hull not getting back to full strenght during in-flight repairs, simply because in-flight repairs cannot achieve the same quality as repairs done in a shipyard.

Terms like `"hull repair ineffectiveness"` or `"hull repair limit"` feel more appropriate to communicate this, but as said; English isn't my native language, so if native speakers tell me that "wear" is appropriate here, then I don't have objections to the word wear being used here.

> > Let the mechanic work on repaired hull points, so an efficiency of 0.9 would mean, for every 10 points of hull restored, the maximum moves down by 1. (Not based on frames, casual players don't know frames.)
> 
> The way it's set up is designed to be intuitive with existing generators, which work with units per frame. Players can see 10 hull wear / second and understand that they'll lose that much max HP while their hull repair outfits are running, the same way they can understand their energy consumption. I'd rather not connect the repair rate to the speed of other repair outfits with an efficiency type attribute, because that reduces the number of independent outfit archetypes we can create and produces complications when stacking multiple outfits. For example, how would your 'wear efficiency' change if you installed two outfits with 0.5 repair efficiency? At the moment all we can do is add attributes together, but it wouldn't make sense for them together to produce 100% efficient hull repair.

Okay, consistency is valuable and a good reason to stick to frames. And you are right, my example should be "inefficiency" or even better "ineffectiveness" to make sure that stacking has the desired effect.

File: data/_ui/tooltips.txt (hunk 5 above)
[the remark this answers is not in the document]
Comment (petervdmeer) [reply data/_ui/tooltips.txt:286]: Shorter is indeed better.

**Comment by tibetiroka** [comment]
@Azure3141 Can you fix the merge conflicts? I also see some unresolved comments above, do you know if you can get around to them?

**Pull request closed without merging**